### PR TITLE
Replace GTS with CDT

### PR DESCRIPTION
**Pull request by mjcarroll**
This removes the dependency on GTS.

* It removes the Mesh boolean operations functions, which were part of the building editor in Gazebo-classic but are unused in modern gazebo.
* It replaces the GTS delaunay triangulation with the header-only CDT library.

### CODE REVIEW

**Review by scpeters — COMMENTED** [review]
some small whitespace nits

and I've prepared a change for homebrew-simulation for after this lands: https://github.com/osrf/homebrew-simulation/pull/2704

**Review by azeey — COMMENTED** [review]
The MeshCSG code was used in Gazebo-classic to implement the Building Editor. Since there is no plan to port this to modern Gazebo (we can potentially use [rmf_site](https://github.com/open-rmf/rmf_site) for this in the future), I'd say let's go ahead and remove it from gz-common.

**Comment by iche033** [comment]
gts dep removal in the gz-common6-release repo when this lands: https://github.com/gazebo-release/gz-common6-release/pull/4

**Comment by iche033** [comment]
I tested this with the polylines.sdf world and I think some of the triangles may be oriented incorrectly, i.e. normals are in the opposite direction:

left: GTS vs right: CDT

cdt_tri

File: graphics/src/CDT/CDT.h
[the remark this answers is not in the document]
Comment (scpeters) [reply graphics/src/CDT/CDT.h:1]: 👍

File: graphics/include/gz/common/MeshManager.hh [hunk 1]
```diff
@@ -270,16 +270,6 @@ namespace gz
...
 
-      /// \brief Create a boolean mesh from two meshes
-      /// \param[in] _name the name of the new mesh
-      /// \param[in] _m1 the parent mesh in the boolean operation
-      /// \param[in] _m2 the child mesh in the boolean operation
-      /// \param[in] _operation the boolean operation applied to the two meshes
-      /// \param[in] _offset _m2's pose offset from _m1
-      public: void CreateBoolean(const std::string &_name, const Mesh *_m1,
-          const Mesh *_m2, const int _operation,
-          const gz::math::Pose3d &_offset = gz::math::Pose3d::Zero);
```
Comment (scpeters) [inline graphics/include/gz/common/MeshManager.hh:281]: * [ ] update the migration guide to note the dependency change and the removal of this API

**Comment by scpeters** [comment]
> I tested this with the polylines.sdf world and I think some of the triangles may be oriented incorrectly, i.e. normals are in the opposite direction:

I added a printout to the DelaunayTriangulation test to show the vertices for the triangles:

## patch

<details>

~~~
diff --git a/graphics/src/DelaunayTriangulation_TEST.cc b/graphics/src/DelaunayTriangulation_TEST.cc
index df919d3..c441257 100644
--- a/graphics/src/DelaunayTriangulation_TEST.cc
+++ b/graphics/src/DelaunayTriangulation_TEST.cc
@@ -69,6 +69,32 @@ TEST_F(DelaunayTriangulation, DelaunayTriangulation)
   // same as number of vertices in the path
   EXPECT_EQ(subMesh.VertexCount(), 8u);
 
+  std::cerr << subMesh.Max() << '\n';
+  std::cerr << subMesh.Min() << '\n';
+  // Print vertices
+  std::cerr << "vertices: \n";
+  for (int v = 0; v < subMesh.VertexCount(); ++v)
+  {
+    std::cerr << subMesh.Vertex(v) << '\n';
+  }
+  // Print normals
+  std::cerr << "normals: \n";
+  for (int v = 0; v < subMesh.NormalCount(); ++v)
+  {
+    std::cerr << subMesh.Normal(v) << '\n';
+  }
+
   // there should be 8 triangles.
   EXPECT_EQ(subMesh.IndexCount() / 3u, 8u);
+  // Print triangles
+  std::cerr << "triangles: \n";
+  for (int i = 0; i < subMesh.IndexCount() / 3u; ++i)
+  {
+    int i1 = subMesh.Index(i*3 + 0);
+    int i2 = subMesh.Index(i*3 + 1);
+    int i3 = subMesh.Index(i*3 + 2);
+    std::cerr << '[' << subMesh.Vertex(i1) << "]\n"
+              << '[' << subMesh.Vertex(i2) << "]\n"
+              << '[' << subMesh.Vertex(i3) << "]\n\n";
+  }
 }
~~~

</details>

## output with GTS

<details>

~~~
[ RUN      ] GTSMeshUtils.DelaunayTriangulation
1 1 0
0 0 0
vertices: 
0.75 0.75 0
1 0 0
1 1 0
0.75 0.25 0
0.25 0.75 0
0 1 0
0.25 0.25 0
0 0 0
normals: 
triangles: 
[0.75 0.75 0]
[1 1 0]
[1 0 0]

[0.75 0.25 0]
[0.75 0.75 0]
[1 0 0]

[0.25 0.75 0]
[0.25 0.25 0]
[0 1 0]

[0.75 0.25 0]
[1 0 0]
[0.25 0.25 0]

[1 0 0]
[0 0 0]
[0.25 0.25 0]

[0.25 0.25 0]
[0 0 0]
[0 1 0]

[0.75 0.75 0]
[0.25 0.75 0]
[1 1 0]

[0.25 0.75 0]
[0 1 0]
[1 1 0]

[       OK ] GTSMeshUtils.DelaunayTriangulation (1 ms)
~~~

</details>

## output with CDT

<details>

~~~
[ RUN      ] DelaunayTriangulation.DelaunayTriangulation
1 1 0
0 0 0
vertices: 
0 0 0
1 0 0
1 1 0
0 1 0
0.25 0.25 0
0.25 0.75 0
0.75 0.75 0
0.75 0.25 0
normals: 
triangles: 
[0.75 0.75 0]
[1 1 0]
[0 1 0]

[0.25 0.75 0]
[0.75 0.75 0]
[0 1 0]

[0.25 0.25 0]
[0 1 0]
[0 0 0]

[0.75 0.25 0]
[0.25 0.25 0]
[1 0 0]

[0.25 0.75 0]
[0 1 0]
[0.25 0.25 0]

[0.75 0.25 0]
[1 0 0]
[0.75 0.75 0]

[1 0 0]
[1 1 0]
[0.75 0.75 0]

[1 0 0]
[0.25 0.25 0]
[0 0 0]

[       OK ] DelaunayTriangulation.DelaunayTriangulation (1 ms)
~~~

</details>

**Comment by iche033** [comment]
> > I tested this with the polylines.sdf world and I think some of the triangles may be oriented incorrectly, i.e. normals are in the opposite direction:
> 
> I added a printout to the DelaunayTriangulation test to show the vertices for the triangles:
> 
> ## patch
> ## output with GTS
> ## output with CDT

Thanks, I created a fix for this in #623. It also addresses other feedback in this PR.

File: graphics/src/DelaunayTriangulation_TEST.cc [hunk 2]
```diff
@@ -63,18 +57,39 @@ TEST_F(GTSMeshUtils, DelaunayTriangulation)
...
   EXPECT_TRUE(result);
 
   // same as number of vertices in the path
   EXPECT_EQ(subMesh.VertexCount(), 8u);
 
   // there should be 8 triangles.
   EXPECT_EQ(subMesh.IndexCount() / 3u, 8u);
+
+  // verify that triangles have clockwise winding
+  for (int t = 0; t < subMesh.IndexCount() / 3u; ++t)
```
Comment (scpeters) [inline graphics/src/DelaunayTriangulation_TEST.cc:76]:
```suggestion
  for (unsigned int t = 0; t < subMesh.IndexCount() / 3u; ++t)
```
apologies I missed the compiler warning in #624

**Comment by scpeters** [comment]
I think we could ignore `graphics/src/CDT` in [codecov.yml](https://github.com/gazebosim/gz-common/blob/mjcarroll/cdt/codecov.yml)

**Comment by iche033** [comment]
> I think we could ignore `graphics/src/CDT` in [codecov.yml](https://github.com/gazebosim/gz-common/blob/mjcarroll/cdt/codecov.yml)

I updated the ignore files in [7ddc32a](https://github.com/gazebosim/gz-common/pull/617/commits/7ddc32acca654f93fd9c17c12b7124c491459105)

**Review by scpeters — APPROVED** [review]

**Review by scpeters — CHANGES_REQUESTED** [review]
forgot one thing: we can remove the following line

https://github.com/gazebosim/gz-common/blob/mjcarroll/cdt/.github/ci/packages.apt#L8

**Review by scpeters — APPROVED** [review]

**Review by iche033 — APPROVED** [review]

**Comment by iche033** [comment]
removed gts dep from package.xml in a159b79

**Pull request merged**